src/core/net/tcp6.cpp

```diff
@@ -136,20 +157,121 @@ Error Tcp::Endpoint::Abort(void)
+
+    Tcp &tcp = GetInstance().Get<Tcp>();
+    SuccessOrExit(error = tcp.mEndpoints.Find(*this, prev));
+    tcp.mEndpoints.PopAfter(prev);
+    SetNext(nullptr);
+
+exit:
+    return error;
+}
+
+uint8_t Tcp::Endpoint::TimerFlagToIndex(uint8_t aTimerFlag)
+{
+    OT_UNUSED_VARIABLE(aTimerFlag);
+    /*
+     * TODO: Convert from the timer flag provided by TCPlp to the index in
+     * our timers array.
+     */
+    return 0;
+}
+
+bool Tcp::Endpoint::IsTimerActive(uint8_t aTimerIndex)
+{
+    OT_UNUSED_VARIABLE(aTimerIndex);
+    /* TODO: Check whether TCPlp has marked this timer as active. */
+    return false;
+}
+
+void Tcp::Endpoint::SetTimer(uint8_t aTimerFlag, uint32_t aDelay)
+{
+    TimeMilli now         = TimerMilli::GetNow();
+    TimeMilli newFireTime = now + aDelay;
+    uint8_t   timerIndex  = TimerFlagToIndex(aTimerFlag);
+
+    mTimers[timerIndex] = newFireTime.GetValue();
+    otLogDebgTcp("Set timer %u with delay %u", static_cast<unsigned int>(timerIndex),
+                 static_cast<unsigned int>(aDelay));
+
+    GetInstance().Get<Tcp>().ResetTimer(false);
+}
+
+void Tcp::Endpoint::CancelTimer(uint8_t aTimerFlag)
+{
+    /*
+     * TCPlp has already cleared the timer flag before calling this, so all
+     * that we have to do is stop the actual timer if needed.
+     */
+    if (GetInstance().Get<Tcp>().mTimer.GetFireTime() == TimeMilli(mTimers[TimerFlagToIndex(aTimerFlag)]))
+    {
+        GetInstance().Get<Tcp>().ResetTimer(false);
+    }
+}
+
+void Tcp::Endpoint::FirePendingTimers(TimeMilli aNow)
+{
+    /*
+     * NOTE: Firing a timer might potentially activate/deactivate other timers.
+     * If timers x and y expire at the same time, but the callback for timer x
+     * (for x < y) cancels or postpones timer y, should timer y's callback be
+     * called? Our answer is no, since timer x's callback has updated the
+     * TCP stack's state in such a way that it no longer expects timer y's
+     * callback to to be called. Because the TCP stack thinks that timer y
+     * has been cancelled, calling timer y's callback could potentially cause
+     * problems.
+     */
+    for (uint8_t timerIndex = 0; timerIndex != kNumTimers; timerIndex++)
+    {
+        if (IsTimerActive(timerIndex))
+        {
+            Time expiry(mTimers[timerIndex]);
+            if (expiry <= aNow)
+            {
+                /* TODO: Call TCPlp's callback for this timer. */
+            }
+        }
+    }
+}
+
+bool Tcp::Endpoint::GetEarliestActiveTimer(TimeMilli aNow, uint32_t &aDelayUntilNextActiveTimer)
+{
+    bool hasUnfiredTimer       = false;
+    aDelayUntilNextActiveTimer = Timer::kMaxDelay;
+
+    for (uint8_t timerIndex = 0; timerIndex != kNumTimers; timerIndex++)
+    {
+        if (IsTimerActive(timerIndex))
+        {
+            Time expiry(mTimers[timerIndex]);
+            hasUnfiredTimer            = true;
+            aDelayUntilNextActiveTimer = OT_MIN(aDelayUntilNextActiveTimer, expiry - aNow);
+        }
+    }
```

Comment:
There are some helper methods in `Time` class which may help here.

The `Time` class overloads the `operator<()` to handle the comparison even when actual time may roll-over (this is the main purpose of it 😃...)

We also have `now.GetDistantFuture()` and `now.GetDistantPast()` which we can then use to safely compare against a bunch of other `Time` instances to find the earliest or latest time.

As example, you can check `Commissioner::UpdateJoinerExpirationTimer()` in `commissioner.cpp` on how they are used (or `Client::HandleTimer()` in `dns_clinet.cpp`).

Reply:
There is also `TimerMilli::FireAtIfEarlier()` which is helpful in many situations (in addition to simpler `TimerMilli::FireAt()`).

```c++
   /**
     * This method (re-)schedules the timer with a given a fire time only if the timer is not running or the new given
     * fire time is earlier than the current fire time.
     *
     * @param[in]  aFireTime  The fire time.
     *
     */
    void FireAtIfEarlier(TimeMilli aFireTime);
```

Reply:
Thanks. I'm aware of `Time` and `TimeMilli`. But in this case, I'm using `OT_MIN` to compare delays (delay = difference between two times), not times. Looking at the time library, it looks like delays are represented as `uint32_t` rather than a dedicated class, and comparing them is done using the default `<` operator rather than an overloaded one.

I suppose I can re-work this code to just compute the `Time` of expiry rather than a delay starting at the current time. Would that be preferable?

Reply:
I went ahead and refactored this to work with absolute times instead of delays and use `FireAt`/`FireAtIfEarlier`.